src/audio/mic_privacy_manager/mic_privacy_manager.c

```diff
@@ -0,0 +1,275 @@
+
+	if (mic_privacy_policy == FW_MANAGED)
+	{
+		LOG_INF("mic_privacy FW_MANAGED");
+
+		mic_privacy_api->set_fw_managed_mode(true);
+	}
+
+	//Enable interrupts
+	if (mic_privacy_policy == HW_MANAGED) {
+		LOG_INF("mic_privacy init HW_MANAGED");
+		enable_dmic_irq(true);
+		}
+
+	else {
+		enable_fw_managed_irq(true);
+	}
+
+	return 0;
+}
+
+int mic_privacy_manager_get_policy(void)
+{
+	LOG_INF("mic_privacy_manager_get_policy ");
+
+	mic_privacy_api = (struct mic_privacy_api_funcs *) mic_priv_dev->api;
+
+	return mic_privacy_api->get_policy();
+}
+
+uint32_t mic_privacy_get_policy_register(void)
+{
+	if (mic_priv_dev == NULL)
+		return -EINVAL;
+	mic_privacy_api = (struct mic_privacy_api_funcs *) mic_priv_dev->api;
+
+	return mic_privacy_api->get_privacy_policy_register_raw_value();
+}
+
+void enable_fw_managed_irq(bool enable_irq)
+{
+	LOG_INF("enable_fw_managed_irq %d", enable_irq);
+
+	if (enable_irq)
+		mic_privacy_api->enable_fw_managed_irq(true, handle_fw_managed_interrupt);
+	else
+		mic_privacy_api->enable_fw_managed_irq(false, NULL);
+}
+
+void handle_dmic_interrupt(void *const self, int a, int b)
+{
+    LOG_INF("mic_privacy handle_dmic_interrupt");
+
+    if (mic_privacy_api->get_dmic_irq_status() )
+    {
+        uint32_t mic_disable_status = mic_privacy_api->get_dmic_mic_disable_status();
+        struct mic_privacy_settings settings = fill_mic_priv_settings(mic_disable_status);
+        propagate_privacy_settings(&settings);
+        mic_privacy_api->clear_dmic_irq_status();
+    }
+}
+
+
+void enable_dmic_irq(bool enable_irq)
+{
+    if (enable_irq)
+        mic_privacy_api->enable_dmic_irq(true, handle_dmic_interrupt);
+    else
+        mic_privacy_api->enable_dmic_irq(false, NULL);
+}
+
+
+void handle_fw_managed_interrupt(void * const dev)
+{
+	LOG_INF("handle_fw_managed_interrupt");
+	uint32_t mic_disable_status = mic_privacy_api->get_fw_managed_mic_disable_status();
+
+
+	struct mic_privacy_settings mic_privacy_settings =
+					fill_mic_priv_settings(mic_disable_status);
+
+	propagate_privacy_settings(&mic_privacy_settings);
+
+
+	if (mic_disable_status)
+		mic_privacy_api->set_fw_mic_disable_status(true);
+	else
+		mic_privacy_api->set_fw_mic_disable_status(false);
+
+	mic_privacy_api->clear_fw_managed_irq();
+}
+
+void propagate_privacy_settings(struct mic_privacy_settings *mic_privacy_settings)
+{
+	LOG_INF("propagate_privacy_settings");
+	notifier_event(mic_priv_dev, NOTIFIER_ID_MIC_PRIVACY_STATE_CHANGE,
+		NOTIFIER_TARGET_CORE_ALL_MASK,
+		mic_privacy_settings, sizeof(struct mic_privacy_settings));
+}
+
+uint32_t get_dma_zeroing_wait_time(void)
+{
+	return (uint32_t)mic_privacy_api->get_dma_data_zeroing_wait_time();
+}
+
+
+uint32_t get_privacy_mask(void)
+{
+	if (mic_privacy_policy == HW_MANAGED)
+	{
+		return mic_privacy_api->get_dma_data_zeroing_link_select();
+	}
+	//hardcoded for FW_MANAGED
+	return 0xFFFFFFFF;
+}
+
+struct mic_privacy_settings fill_mic_priv_settings(uint32_t mic_disable_status)
```

Comment:
I think supplying a structure to be filled as a parameter is preferred over returning one